airflow/providers/google/cloud/hooks/dataflow.py

```diff
@@ -420,7 +420,9 @@ def _check_dataflow_job_state(self, job) -> bool:
                     "JOB_STATE_DRAINED while it is a batch job"
                 )
 
-        if not self._wait_until_finished and current_state == self._expected_terminal_state:
+        if current_state == self._expected_terminal_state:
+            if self._expected_terminal_state == DataflowJobStatus.JOB_STATE_RUNNING:
+                return not self._wait_until_finished
```

Comment:
I never used this operator, but I wonder what is `_expected_terminal_state` and why we may return False if `_wait_until_finished` is set the True and the state is equals to the provided expected terminal state :thinking:

Reply:
`_expected_terminal_state` is optional parameter indicating which state is considered as a success (discussed in https://github.com/apache/airflow/pull/8553#discussion_r416136954). I think the coexistence of two similar parameters, `_wait_until_finished` and `_expected_terminal_state`, causes this weird implementation (What `_wait_until_finished` is True has the same meaning as what `_expected_terminal_state` is DataflowJobStatus.JOB_STATE_DONE). So, I think it would be better to remove `_wait_until_finished` if we adopt `_expected_terminal_state`.

Reply:
> why we may return False if _wait_until_finished is set the True and the state is equals to the provided expected terminal state

This is because `_wait_until_finished` was applied preferentially compared to `_expected_terminal_state` in the original implementation.

Reply:
@asafeunico As the temporary solution, you can use the version 10.8.0 of apache-airflow-providers-google without this problem.

Reply:
I'm the author of #34217 and sincerely apologize for the mess!
I ensured that the existing tests passed, but I didn't consider adding the `JOB_STATE_DONE` - which is problematic for that `if` case.
As @champon1020 stated, `_wait_until_finished` should have been applied preferentially compared to `expected_terminal_state` to avoid breaking default behavior.
I agree that having both params is somewhat confusing, and we should revise deprecating `_wait_until_finished` in the future.
The suggested solution LGTM, thank you for fixing it up!